### PR TITLE
Update Config module with displaying dynamic values in with patternProperties in schema

### PR DESCRIPTION
**Pull request by Pe5h4**
ASAB - Config refactorisation with patternProperties in schema

### CODE REVIEW

**Comment by Pe5h4** [comment]
@ateska This is the first iteration of refactorization of ASAB-Config module based on friday's discussion and outputs

If there will be anything broader to add/refactor from you or something unclear, we can have a short call about it.

Values to render and section (pattern)properties are stored in formStruct in following flow described in points:

After data from endpoints are obtained:
 1] add sections and section keys from properties to formStruct with appropriate data (if properties present in the schema)
https://github.com/TeskaLabs/asab-webui/blob/a8d7e819d99300392a405b53ad36528bb3634841/src/modules/asab-config/ConfigContainers/ConfigEditor.js#L109
 https://github.com/TeskaLabs/asab-webui/blob/a8d7e819d99300392a405b53ad36528bb3634841/src/modules/asab-config/ConfigContainers/ConfigEditor.js#L115
 2] add sections and section keys from patternProperties to formStruct with appropriate data (if patternProperties sections are present in the schema)
https://github.com/TeskaLabs/asab-webui/blob/a8d7e819d99300392a405b53ad36528bb3634841/src/modules/asab-config/ConfigContainers/ConfigEditor.js#L144
https://github.com/TeskaLabs/asab-webui/blob/a8d7e819d99300392a405b53ad36528bb3634841/src/modules/asab-config/ConfigContainers/ConfigEditor.js#L148

https://github.com/TeskaLabs/asab-webui/blob/a8d7e819d99300392a405b53ad36528bb3634841/src/modules/asab-config/ConfigContainers/ConfigEditor.js#L176-L181

 3] add rest of the values to adHoc
https://github.com/TeskaLabs/asab-webui/blob/a8d7e819d99300392a405b53ad36528bb3634841/src/modules/asab-config/ConfigContainers/ConfigEditor.js#L171-L174
 4] set values from formStruct and from adHoc for register (with the usage of `setValue` comonent of react-hook-form)
https://github.com/TeskaLabs/asab-webui/blob/a8d7e819d99300392a405b53ad36528bb3634841/src/modules/asab-config/ConfigContainers/ConfigEditor.js#L186-L201
 5] render input forms and data according to formStruct properties
https://github.com/TeskaLabs/asab-webui/blob/a8d7e819d99300392a405b53ad36528bb3634841/src/modules/asab-config/ConfigContainers/ConfigEditor.js#L307-L328

Sections with adHoc cannot be saved now (it does not fit the schema, service is throwing errors) and support for patternProperties will be added as well (now we cannot save values to configs with patternProperties) - I discussed this with Mithun and he is going to work on that next sprint.

Here is the flow with examples of configuration with adHoc sections or/and adHoc keys (values). On gif image there is also an example of schemas with patternProperties (and as well with adHocs):

config-refactorisation

I have also question - should we support nested patternProperties (meant for keys)? I suppose for our `section-key-value` configs this wont be necessary. Here is the example:

```
...
 "patternProperties": {
        "^section:source:(?<varname>.*)$": {
            "type": "object",
            "title": "Section configuration",
            "description": "Secition source",
            "default": {},
            "examples": [
                {
                    "^section:source:name:(?<keyname>.*)$": "My Key"
                }
            ],
            "required": [
                "^section:source:name:(?<keyname>.*)$"
            ],
            "patternProperties": {
                "^section:source:name:(?<keyname>.*)$": {
                    "type": "string",
                    "title": "Name",
                    "description": "Fill the name of the Key",
                    "default": "",
                    "examples": [
                        "MyOwnKey"
                    ]
                }
            },
            "additionalProperties": false
        },
...
```

**Comment by ateska** [comment]
Support for the `pattern on key` will be added eventually (if ever).

File: src/modules/asab-config/ConfigContainers/ConfigEditor.js [hunk 1]
```diff
@@ -22,113 +22,181 @@ import {
...
-					let v = {};
-					if (type?.properties[section] == undefined) {
-						s[key] = values[section][key];
-						arrSections.push(s);
-						ahSections[section] = arrSections;
-					} else if (type?.properties[section]?.properties[key] == undefined) {
-						v[key] = values[section][key];
-						arrValues.push(v);
-						ahValues[section] = arrValues;
+		let fs = {};
```
Comment (ateska) [inline src/modules/asab-config/ConfigContainers/ConfigEditor.js:100]: Here we need more descriptive variable name.

File: src/modules/asab-config/ConfigContainers/ConfigEditor.js [hunk 2]
```diff
@@ -22,113 +22,181 @@ import {
...
+		let schemaProps = {};
+		let data = {};
```
Comment (ateska) [inline src/modules/asab-config/ConfigContainers/ConfigEditor.js:102]: Here we need more descriptive variable name, `data` doesn't reveal too much.

**Review by ateska — CHANGES_REQUESTED** [review]
Few remarks but this is a brilliant direction.

**Comment by Pe5h4** [comment]
@ateska Updates:
- variable names for formStruct renamed https://github.com/TeskaLabs/asab-webui/blob/4ab015a9c5c620246a068dcb7ce2746f569c6b9d/src/modules/asab-config/ConfigContainers/ConfigEditor.js#L98-L100
- feature to handle cases, when key-value is not present in the configuration, but schema is present for this case (it can be the case of an empty config) https://github.com/TeskaLabs/asab-webui/blob/4ab015a9c5c620246a068dcb7ce2746f569c6b9d/src/modules/asab-config/ConfigContainers/ConfigEditor.js#L118-L122 - empty string is added to the section-key (we cannot use undefined value since it would not be rendered properly, but we can use empty string for every input type [number, boolean, etc.])
- implemented function to convert values based on value types on submit. onSubmit function converts all the updated values to string, so it has to be parsed back by the appropriate `type` before data are being processed to the service (as we discussed yesterday with you and Mithun) https://github.com/TeskaLabs/asab-webui/blob/4ab015a9c5c620246a068dcb7ce2746f569c6b9d/src/modules/asab-config/ConfigContainers/ConfigEditor.js#L313
- I commented editing values from JSON viewer (so now it is disabled), since you can specify the type of the value from JSON viewer and if you choose the wrong one, the configuration will not be saved. Anyway from what I remember, we were speaking of removing the JSON view completely in the future.
- other minor updates as seen from the commits

When Mithun will provide functional update on asab-config service, I will add a new gif image with the flow

**Comment by ateska** [comment]
Looks good to me so far.

**Comment by Pe5h4** [comment]
@ateska In addition to above updates you have already aware of, I extended documentation for asab-config module (schema and config examples, supported input types) https://github.com/TeskaLabs/asab-webui/blob/da080a3fb044fdfe07dd33b922b36be40e305ad4/doc/asab-config.md

The creation of new (empty) configuration from asab-config module and sorting sections in the tree menu by name (as we spoke of) will be done in the next iteration

Here is the read/save flow for properties and pattern properties also with ad hoc values and ad hoc sections:
pattern-props-asab-config

**Comment by ateska** [comment]
Why there is `config` in  `/config/Discover` path in the URL? ... that seems to be dangerous to me.

**Comment by ateska** [comment]
Also word "key" in the config item is a bit redundant already.

**Comment by ateska** [comment]
Tenant value in a config needs to be removed.

Let's have a review workshop for this, there are few additional remarks I have.

**Review by ateska — APPROVED** [review]

**Comment by Pe5h4** [comment]
> Why there is `config` in `/config/Discover` path in the URL? ... that seems to be dangerous to me.

@ateska This is a path to the config section. We decided to have path URL as `config/sectionName/configName`, (default is `config/$/$`) so the `config/Discover` just points to the section.

**Comment by Pe5h4** [comment]
> Also word "key" in the config item is a bit redundant already.

@ateska Yes, this is for the purpose of exercise. The name or the description is taken from the schema. Anyway, we should have a workshop about how to properly configure schemas (I suppose also together with Mithun)

Anyway, lets have a talk about it on the review, I will merge it only if there will be only after the review and only if there will be minors, which can be done in the next iteration.

**Pull request merged**